### PR TITLE
Closes #2336 fix bug in derive_param_tte

### LINKED ISSUE #2336: Bug: `derive_param_tte()` fails if `dataset` and `by_vars` is specified
### What happened?

 `derive_param_tte()` fails if `dataset` and `by_vars` is specified because the check on duplicate parameter codes fails.

### Session Information

_No response_

### Reproducible Example

```
library(dplyr)
library(lubridate)
library(admiral)
adsl <- tibble::tribble(
  ~USUBJID, ~TRTSDT,           ~TRTEDT,           ~EOSDT,
  "01",     ymd("2020-12-06"), ymd("2021-03-02"), ymd("2021-03-06"),
  "02",     ymd("2021-01-16"), ymd("2021-01-20"), ymd("2021-02-03")
) %>%
  mutate(STUDYID = "AB42")

ae <- tibble::tribble(
  ~USUBJID, ~AESTDTC,     ~AESEQ, ~AEDECOD,
  "01",     "2021-01-03",      1, "Flu",
  "01",     "2021-03-04",      2, "Cough",
  "01",     "2021-01-01",      3, "Flu"
) %>%
  mutate(
    STUDYID = "AB42",
    AESTDT = ymd(AESTDTC)
  )

ttae <- event_source(
  dataset_name = "ae",
  date = AESTDT,
  set_values_to = exprs(
    EVENTDESC = "AE",
    SRCDOM = "AE",
    SRCVAR = "AESTDTC",
    SRCSEQ = AESEQ
  )
)

eot <- censor_source(
  dataset_name = "adsl",
  date = pmin(TRTEDT + days(10), EOSDT),
  censor = 1,
  set_values_to = exprs(
    EVENTDESC = "END OF TRT",
    SRCDOM = "ADSL",
    SRCVAR = "TRTEDT"
  )
)

  derive_param_tte(
    dataset = adsl %>% select(STUDYID, USUBJID) %>% mutate(PARAMCD = "XYZ"),
    dataset_adsl = adsl,
    by_vars = exprs(AEDECOD),
    start_date = TRTSDT,
    event_conditions = list(ttae),
    censor_conditions = list(eot),
    source_datasets = list(adsl = adsl, ae = ae),
    set_values_to = exprs(
      PARAMCD = paste0("TTAE", as.numeric(as.factor(AEDECOD))),
      PARAM = paste("Time to First", AEDECOD, "Adverse Event"),
      PARCAT1 = "TTAE",
      PARCAT2 = AEDECOD
    )
  )
```
results in
```
Error in match(x, table, nomatch = 0L) : 
  'match' requires vector arguments
```

### PR DESCRIPTION
**Pull request by millerg23**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF version of it to the same folder.
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# admiral 1.0.2
+
+- Fix bug in `derive_param_tte()` where argument `dataset` populated and argument `by_vars` is dynamic. (#2336)
```
Comment (bms63) [inline NEWS.md:3]: I thought the issue was due to duplicated information being detected?

File: tests/testthat/test-derive_param_tte.R [hunk 2]
```diff
@@ -805,33 +805,135 @@ test_that("derive_param_tte Test 11: ensuring ADT is not NA because of missing s
   )
 })
 
+## Test 12: test dataset and dynamic byvars populated ----
```
[the remark this answers is not in the document]
Comment (millerg23) [reply tests/testthat/test-derive_param_tte.R:808]: Apologies, its when the PARAMCD in SET_VALUES_TO is not static, not `BY_VARS`. I will re-word. See new unit test, the PARAMCD is an expression, not a constant value.

File: R/derive_param_tte.R [hunk 3]
```diff
@@ -450,6 +447,13 @@ derive_param_tte <- function(dataset = NULL,
     new_param <- select(new_param, !!!negate_vars(by_vars))
   }
 
+  # check newly created parameter(s) do not already exist
+  if (!is.null(set_values_to$PARAMCD) && !is.null(dataset)) {
+    for (i in seq_along(new_param$PARAMCD)) {
```
Comment (bundfussr) [inline R/derive_param_tte.R:452]: I would restrict `new_params$PARAMCD` to unique values (`unique_params <- unique(new_params$PARAMCD)`). Otherwise each `PARAMCD` value is checked several times.

File: NEWS.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
+# admiral 1.0.2
+
+- Fix bug in `derive_param_tte()` where argument `dataset` populated and `PARAMCD` in `SET_VALUE_TO` argument is an expression. (#2336)
```
Comment (bundfussr) [inline NEWS.md:3]:
```suggestion
- Fix bug in `derive_param_tte()` where argument `dataset` populated and `PARAMCD` in `set_values_to` argument is an expression. (#2336)
```

File: R/derive_param_tte.R (hunk 3 above)
Comment (millerg23) [reply R/derive_param_tte.R]: Done.

**Review by bundfussr — APPROVED** [review]

File: NEWS.md [hunk 5]
```diff
@@ -1,3 +1,7 @@
+# admiral 1.0.2
+
+- Fix bug in `derive_param_tte()` where argument `dataset` populated and `PARAMCD` in `set_values_to` argument is an expression. (#2336)
```
Comment (bms63) [inline NEWS.md:3]: Hey Gordon - I guess I am still confused about what we are telling the user here, what we are fixing and there might be a dose of me not fully following  :)

The bug is from the function picking up potential false positives around the parameters and falling over with the duplication error?

The fix is now we check to make sure the parameters do not already exist in the datasets and give feedback to the user?

I feel like reading the issue and news bullet has only confused me!!

Comment (millerg23) [reply NEWS.md:3]: Previously the check was looking at the contents of what `PARAMCD` was in `set_values_to` argument, and seeing if the `PARAMCD` already existed in dataset that is passed into `dataset` argument.  If `PARAMCD` was not an expression ie `PARAMCD = "XYZ"` then check worked. However, check was very early in function, so if `PARAMCD` to be created was an expression, and wasn't resolved yet, this caused the ERROR. I moved the check to near the end, where `PARAMCD` is resolved in the dataset holding the new params. Happy to change the wording in the NEWS item!

Comment (bms63) [reply NEWS.md:3]: ooohhh could that be the new summary just a little more polished please.

thanks again for digging into this and ELI5 to me!!

**Review by bms63 — APPROVED** [review]
Thanks @millerg23! 

@bundfussr @rossfarrugia @manciniedoardo 

 Is it okay to deliver to CRAN on March 4th.  That way we have a full month between releases?

image
